### PR TITLE
Quest Spec

### PR DESCRIPTION
**Pull request by LegitDongo**
Hey guys 👋 

Here's the rough draft for quests that I've come up with

I tried to keep a lot of it relatively based, and where some specifics were needed, I tried to keep things based off already existing parameters (for monster fields, use the same naming/typing(if applicable) of already existing hooks so that things are mostly consistent

I didn't really want to include the item list, since it would add another list for webhook endpoints to maintain, but it feels necessary for when you're trying to filter based on the IDs only, and string conditions aren't really fun.

I'm sure I've missed something, or something isn't really needed, so *please* review, and I will revise if most agree.

Sorry for the delay, and thanks for staying tuned 😄

### CODE REVIEW

**Comment by sn0opy** [comment]
Talked to @LegitDongo about this on Discord already. I wouldn't use the IDs from the protos, since they CAN change in the future. Sticking to their name representation is probably more future proof and human readable.

**Comment by kvangent** [comment]
Strings get tricky to compare, particularly if they change over time. Personally I think using the proto id's for a baseline are OK, so ong as it's clear that they won't change just because the proto's do.

File: quests.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
+#Root Quest
```
Comment (kvangent) [inline quests.md]: nit: Drop the root

File: quests.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+| Key           | Type   | Description                         |
+|---------------|--------|-------------------------------------|
+| longitude     | float  |                                     |
+| latitude      | float  |                                     |
+| pokestop_url  | string |                                     |
+| pokestop_name | string |                                     |
+| updated       | int    | epoch time                          |
```
Comment (kvangent) [inline quests.md]: I don't like `updated` as a key - it seems too vague. Also the description doesn't really tell me what it represents. Is it the last time the quest was observed?

File: quests.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+| conditions    | array  | See [conditions](#conditions) table |
+| rewards       | array  | See [rewards](#rewards) table       |
+
+```json
+[
+  {
+    "type": "quest",
+    "message": {
+      "longitude": float,
```
Comment (kvangent) [inline quests.md]: nit: Examples should show an example value instead of the type (so they don't have json validation issues)

Also it might be a reasonable idea to include a full example in it's own section at the top, making it easier for people to copy/paste when they want to test something out

File: quests.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+| Key           | Type   | Description                                                                   |
+|---------------|--------|-------------------------------------------------------------------------------|
+| type          | int    | See [Conditions Type Array](#conditions-type-array)                           |
+| amount        | int    | A generic number representing how many times something must be done or aquired|
+| detail        | object | See [Conditions Detail Sections](#conditions-detail) for further detail       |
+
+```json
+[
+    {
+      "type": int,
```
Comment (kvangent) [inline quests.md]: Personally for me, I feel like `type` has become a somewhat overloaded term between monster types/etc. I would rather see `cond_type` or `condition_type` (but I'm open to feedback from others on this)

File: quests.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+      "amount": int,
+      "detail": {...}
+    }
+]
+```
+
+### Conditions Type Array
+This came directly out of the protos, and the `type` int field for every condition should represent one of these
```
Comment (kvangent) [inline quests.md]: Carrying over the conversation in the main thread with @sn0opy, I don't think the description needs to state it's derived from the proto because even if the proto changes we will likely keep our id's the same.

File: quests.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+Keep in mind that all values will be accepted, but not all will be valid, even values in this list
+```
```
Comment (kvangent) [inline quests.md]: This should be a table, not a proto definition. 

I would also drop `WITH` from the value names.

File: quests.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+### Conditions Detail
+When the basic fields of the condition don't cover all use-cases, reference below
+
+Some possibilities from the conditions type array were left out because they didn't seem useful to have any detail on
+
+#### Pokemon Type (type `1`)
+See [PA's type array](https://github.com/PokeAlarm/PokeAlarm/blob/master/locales/en.json#L1016) for a complete type
+id/name correlation (minus the padded zeroes)
+```json
```
Comment (kvangent) [inline quests.md:163]: Again, I would rather see a table + a full example of what this condition looks like in json formatting

File: quests.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+{
+  "type_id": [int]
```
Comment (kvangent) [inline quests.md]: Another example of type being overloaded - `mon_type_id`?

File: quests.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+}
+```
+
+#### Pokemon Category (type `2`)
+How they define a group of monsters without any particular relation
+```json
+{
+  "pokemon_id": [int],
```
Comment (kvangent) [inline quests.md]: s/pokemon/monster

File: quests.md [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+  "name": string
+}
+```
+* The `name` field is optional, and probably won't ever be filled
```
Comment (kvangent) [inline quests.md]: If it doesn't serve a purpose, let's leave it out

File: quests.md (hunk 8 above)
Comment (kvangent) [inline quests.md]: Is this an array of values? Or just a single int?

File: quests.md [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+|---------------|--------|---------------------------------------------------------------------------------------|
+| type          | int    | See [the reward types array](#rewards-type-array) for options                         |
+| id            | int    | A generic id representing an object, based on `type`                                  |
+| amount        | int    | A generic amount representing how many must be acquired or completed, based on `type` |
+| detail        | object | See [the rewards detail section](#rewards-detail) for options                         |
+
+
+```json
+{
+  "type": int,
```
Comment (kvangent) [inline quests.md]: `reward_type`

File: quests.md [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+  "id": int,
```
Comment (kvangent) [inline quests.md]: `reward_id`

File: quests.md [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+  "amount": int,
+  "detail": {...}
```
Comment (kvangent) [inline quests.md]: Why nest in detail? Why not just have optional fields within the reward object

File: quests.md [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+{
+  "pokemon_id": int
+}
+```
+* ID of `0` will assume the Rare Candy item
+
+
+
+
+## Item array
```
Comment (kvangent) [inline quests.md]: Maybe we should consider moving the ID definitions into a separate folder? 

For example monster id's will likely be used, I'm not sure it make sense to continue to list them in multiple places.

File: quests.md [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+Once again, straight from protos. The int represents the ID
+```
+enum ItemId {
+	ITEM_UNKNOWN = 0;
```
Comment (kvangent) [inline quests.md]: Again, `ITEM` seems redundant.

Also would prefer a table, as with the other proto definitions.

**Review by kvangent — CHANGES_REQUESTED** [review]

File: quests.md (hunk 4 above)
Comment (sn0opy) [reply quests.md]: I'd stick to `type` just because the context defines what kind of "type" it actually is.

File: quests.md (hunk 2 above)
Comment (sn0opy) [reply quests.md]: I'm with @kvangent. Thing is, the quests won't change more than once a day (ignoring events) and there's always just one quest per stop. I'd go with `added` or as already mentioned `observed`. 

In theory you even have to pass a timezone with it due to quests being purged per timezone at around 0:00. Or the receiver has to decide when to purge it based in the stops lat/lng.

**Comment by 123FLO321** [comment]
@sn0opy I don't agree with that.

Working with numbers that always have a clear definition is easier than strings. (And we use it for other stuff as well like weather, forms costume etc.)
And it's not very likely for proto ids to change (they didn't change so far).
Even if they did change it would be easy to re-map the new ids.

**Comment by 123FLO321** [comment]
@LegitDongo We should also define if a value is required or optional.

File: quests.md (hunk 2 above)
Comment (LegitDongo) [reply quests.md]: In the case of community days, quests change 3 times that day.

I know that most people will want their alarm systems to show all of the new quests.

Particularly like with the most recent feebas event.

File: quests.md (hunk 10 above)
Comment (LegitDongo) [reply quests.md]: Afaik, this has never been filled, but may be later

As to what value it will add, I'm not really sure.

File: quests.md (hunk 8 above)
Comment (LegitDongo) [reply quests.md]: an array of ints

File: quests.md (hunk 14 above)
Comment (LegitDongo) [reply quests.md]: I listed this mostly because PA doesn't currently have a game item definition so far

**Comment by Christian-X** [comment]
I'm the author of the Telegram BookOfQuests Bot (so one possible user of this dataset).

RDM is currently using a similar structure. Even as its quite ugly - it makes sense to stay close to the PogoProtos.
Asfar as i can see still missing:

- guid (i think fort_id)
- type (haha.... 3 different types inside that damn struct) the "quest_type" - QuestType in the proto.
- target (how many times the Quest must be accomplished)

**Comment by LegitDongo** [comment]
@Christian-X That was the plan when we were starting this, but we also need things to make sense in a webhook perspective

As per your points:
- Quests are tracked by `pokestop_id`, and are attached to that pokestop for at most 24h, not by gym (fort)
- top level type doesn't really seem necessary since this hook will always be for the quest you get from spinning the pokestop
- this is written a little more in detail in the conditions, unless I'm missing something

Also, if you feel like you'd be using this a lot, feel free to add yourself to the contributors 😄

**Comment by sn0opy** [comment]
@LegitDongo 

> * Quests are tracked by `pokestop_id`, and are attached to that pokestop for at most 24h, not by gym (fort)

I think that's what he meant. Thing is, that the `pokestop_id` is completely missing in your draft ;)

**Comment by kvangent** [comment]
> I'm the author of the Telegram BookOfQuests Bot (so one possible user of this dataset).

Please feel free to add yourself to this repo if you would like :)

> RDM is currently using a similar structure. Even as its quite ugly - it makes sense to stay close to the PogoProtos.

Why? Proto's bring certain benefits with certain costs. By keeping the format but not actually implementing RPC, we are getting all of the costs but none of the gain.

**Comment by kvangent** [comment]
> Working with numbers that always have a clear definition is easier than strings. (And we use it for other stuff as well like weather, forms costume etc.)
+1
> And it's not very likely for proto ids to change (they didn't change so far).
> Even if they did change it would be easy to re-map the new ids.

I think it would be better to try to maintain backwards compatibility wherever possible. Using the proto id's as a baseline is fine, I don't think we should go out of our way to keep them 1:1.

File: quests.md (hunk 14 above)
Comment (kvangent) [reply quests.md]: Right, I'm not saying it's not needed - I'm saying that rather rather than say "this depends on PA's definitions" we should aim for "here are the definitions, PA also implements them".

File: quests.md (hunk 2 above)
Comment (kvangent) [reply quests.md]: I was more concerned that they key string + description should be updated to more clearly reflect what the key adds.

**Comment by Christian-X** [comment]
> As per your points:
> 
>     * Quests are tracked by `pokestop_id`, and are attached to that pokestop for at most 24h, not by gym (fort)
Sorry, yes - i did mean pokestop_id.
>     * top level type doesn't really seem necessary since this hook will always be for the quest you get from spinning the pokestop
>     * this is written a little more in detail in the conditions, unless I'm missing something
No, that's not sufficient. I've spent some time with mapping those quests provided by RDM to my own Questspecification (with started in a handwritten list of observed quests by text, and later where enriched by using RDM protocol information)

Some Quest might be "Send 5 Gifts to friends" - that won't have any condition.
it's got quest_target:5 and quest_type:24 (as per https://github.com/Furtif/POGOProtos/blob/master/src/POGOProtos/Enums/QuestType.proto).

"Land 10 Curveballs" will have something like quest_target:10, quest_type:16 (that only describes "land 10 throws).  additionally, it'll have one condition {as per https://github.com/Furtif/POGOProtos/blob/master/src/POGOProtos/Data/Quests/QuestCondition.proto) - which would have condition_type:15

As far as i can see, you covered how the conditions are used in your condition part of that documentation - but that "base" QuestType is necessary additionally. (As many Quests in Pokemon won't have any condition set. Some others have 2 or 3 . And then there i've seen cases, where a condition is duplicated in some quests "in the wild", but not in others.... (whatever niantic did there)

> Also, if you feel like you'd be using this a lot, feel free to add yourself to the contributors 😄
Seems to make sense...

**Comment by Christian-X** [comment]
Examples for Quest data (from RDM) with no condition, but seperate base-type:

"message" : { "longitude" : 7, "conditions" : [ ], "pokestop_url" : "redacted", "pokestop_id" : "redacted", "type" : 17, "template" : "t1_2019_adventure_buddywalk_hard", "rewards" : [ { "info" : { "shiny" : false, "costume_id" : 0, "form_id" : 0, "pokemon_id" : 73, "gender_id" : 0 }, "type" : 7 } ], "latitude" : 51, "pokestop_name" : "", "target" : 5 },
-- Get 5 Buddy Candies.

{ "_id" : ObjectId("5c4f8cca3fb1ce59d98d8b4f"), "message" : { "target" : 10, "longitude" : 9, "rewards" : [ { "info" : { "item_id" : 1, "amount" : 5 }, "type" : 2 } ], "template" : "challenge_catch_easy", "latitude" : 54, "type" : 4, "pokestop_id" : "xx", "conditions" : [ ], "pokestop_name" : "xx", "pokestop_url" : "" }, "type" : "quest"}
-- Catch 10 Pokemon

{ "_id" : ObjectId("5c4f8cab3fb1ce59d98d89ba"), "message" : { "pokestop_name" : "XX", "pokestop_id" : "", "latitude" : 51, "rewards" : [ { "type" : 3, "info" : { "amount" : 1000 } } ], "target" : 10, "longitude" : 7, "template" : "challenge_hoenn19_catch_specific_plural", "conditions" : [ { "type" : 2, "info" : { "pokemon_ids" : [ 263, 264, 276, 277 ] } } ], "pokestop_url" : "xx", " "type" : 4 }, "type" : "quest" }
-- Catch 10 Pokemon, but only pokedex #263,264,276,277

**Comment by kvangent** [comment]
:+1:  to adding a quest type, seems like a valid argument.

File: quests.md (hunk 2 above)
Comment (KartulUdus) [reply quests.md]: I don't mind the existence of `updated`

It's an optional field that poracle doesn't use

File: quests.md (hunk 13 above)
Comment (KartulUdus) [reply quests.md]: I would really like this to be called `info` as already implemented by RM++ and RDM

**Comment by KartulUdus** [comment]
agree with @123FLO321 
setting minimum required fields would be nice, 
as an example - I've defined minimum required fields to accept a webhook here
https://github.com/KartulUdus/PoracleJS/blob/develop/src/app/schemas.js#L7

**Comment by LegitDongo** [comment]
What do you guys think of the `detail` subsection for conditions and rewards?

Would it be better just to drop it and assume certain keys in the condition or reward object, like we would for the detail section?

So instead of
```json
{
  "type": 7,
  "id": 0,
  "amount": 200,
  "detail": {
    "pokemon_id": 201,
    "form": 15,
    "gender": 3,
    "costume_id": 1
  }
}
```
it would be
```json
{
  "type": 7,
  "id": 0,
  "amount": 200,
  "pokemon_id": 201,
  "form": 15,
  "gender": 3,
  "costume_id": 1
}
```
(or something along those lines, keys not exact)

**Comment by kvangent** [comment]
> Would it be better just to drop it and assume certain keys in the condition or reward object, like we would for the detail section?

+1 to this. The details section doesn't make sense with non-typed maps.

File: quests.md (hunk 9 above)
Comment (LegitDongo) [reply quests.md]: I was basing this off of the monster webhook, since you're getting an encounter with this type

Wouldn't we want the keys the same?

**Comment by KartulUdus** [comment]
this would be more human readable; however, not same format as proto. 

I could rework to handle either, but this should be a question for the current scanner devs

**Comment by fosJoddie** [comment]
I say yes to remove the details field.
Reasoning: 
The proto format is already different than what is being discussed here. Every different condition/reward detail is in fact contained in their own sub section in the protos. This means that in RDM as an example, there's already some processing of these data being done before sending.  Wether this is because of proto interpretation or not I won't go into detail about. 

Point is, since it's already being "processed" in some way, why not make it a bit easier and straight forward to construct. This will also make it easier for anyone wanting to send manually submitted research.

**Comment by 123FLO321** [comment]
In RDM there everything is in info except the type of reward/condition

In my opinion that makes more sense and thats how it’s stored in the protos also. 
@LegitDongo

**Comment by 123FLO321** [comment]
@fosJoedue
RDM only parses the proto and maps it to JSON. There is not really a lot of processing.

**Comment by kvangent** [comment]
> RDM only parses the proto and maps it to JSON. There is not really a lot of processing.

I think Joddie's point is it already has to translate from proto to JSON, the data already undergoes a transformation once. Why not just take the extra step and go with the the more readable structure? I think prioritizing readability is a smart decision - it reduces unnecessary complexity (and as Joddie pointed out mean easier for new folks to understand. particularly when used manually). 

If we follow the proto format now and it changes, we either have to remap the data manually or change the format. I would prefer to change the format as little as possible to maintain backwards compatibility. Since we don't have control over the proto format and can't guarantee it won't change, I don't think it make sense to use it as a hard rule.

**Comment by 123FLO321** [comment]
Making a subarray makes it more readable in my opinion plus we follow the proto format making it easy to adapt proto changes.

**Comment by kvangent** [comment]
>  we follow the proto format making it easy to adapt proto changes.

If changes to the proto mean changes to the webhook format we are proposing here it becomes impossible to guarantee any sort of backwards compatibility. If Project A updates it's outbound format for a proto that's changed, folks that want to use Project A to talk to Project B are out of luck until Project B also updates. It make much more sense to keep a consistent, backwards compatible interface as much as possible to avoid creating a diamond dependency issue.

**Comment by LegitDongo** [comment]
Let me know if I missed anything
The complete changes are in the commit message

Thanks @Christian-X for the examples and the in-depth explanation. The top-level `quest_type` has been added.

The only thing I'm confused about is why we would need the `template` definition, since it's just a string
Likewise, why was `target` attached to the top-level when it's a condition definition?

**Comment by 123FLO321** [comment]
Target is top level in the protos also. 

And template doesn‘t need to be defined (just string) but should be added to the top level.
Could be useful in some cases. 

Maybe even add woker_level to top because rewards can be level dependent. 

Also we should definitely mark if something is required or optional.

**Comment by LegitDongo** [comment]
My vote is to drop `target` since it seems redundant.

I can see why the template can be useful, but imo it's only useful if you are aware of what the string means, and I'd imagine that webhook receivers will not (at least, at the time). The only immediate purpose I can see is determining when there's a quest change on special days, which we do by including the `updated` timestamp. Thus, my vote is to allow this, but not really take any notice of it.

`worker_level` seems to have value, but when have the quests been determined on account level?
Although, I wouldn't put it past them 😄 

I must have forgotten to mark as required/optional on some tables, my bad.

**Comment by theyellow** [comment]
Hi, i'm coding another webhook-based bot (pogorobot) and i'm glad i found this! quest-spec is top, i would like to help if i can.
I'm at beginning of implementation of quests in my bot, so i can't help to decide about structure and types atm, so far i just analyzed json i got from rdm and saved in db..

**Review by kvangent — APPROVED** [review]

**Comment by kvangent** [comment]
@pgo-webhook-collaborators/realdevicemap @pgo-webhook-collaborators/poracle @pgo-webhook-collaborators/map-a-droid - Hey folks, I know everyone is busy with their own projects, but we are eager to finalize this (since it's been WIP for a while and is blocking our official implementation of quests for PA). 

It would be greatly appreciated if everyone could either approve or leave comments with concerns within the next week so we can get it merged. Thanks!

**Comment by kvangent** [comment]
@pgo-webhook-collaborators/map-a-droid @pgo-webhook-collaborators/poracle @pgo-webhook-collaborators/realdevicemap 

Hey everyone,

It would be great if we can wrap this up. In keeping up with the spirit of the repo, it would be ideal if each team could explicitly approve the design so we can move on and start working on others. Please take a look at your nearest convenience.

**Comment by fosJoddie** [comment]
> My vote is to drop `target` since it seems redundant.
> 
> I can see why the template can be useful, but imo it's only useful if you are aware of what the string means, and I'd imagine that webhook receivers will not (at least, at the time). The only immediate purpose I can see is determining when there's a quest change on special days, which we do by including the `updated` timestamp. Thus, my vote is to allow this, but not really take any notice of it.
> 

Leave both target and template, they can be useful for some even if I agree they aren't really "useful" 

> `worker_level` seems to have value, but when have the quests been determined on account level?
> Although, I wouldn't put it past them 😄
> 

Worker level determines rewards for items that are gatelocked, like Max Revives. Quest rewards for a low level worker will become stardust instead and I can see it would be worthwile to note this.

As an independent woman who don't need no man nor my own tools - I can get behind this spec.

**Comment by LegitDongo** [comment]
@fosJoddie Added.
I marked `worker_level` as optional since it isn't required for determining anything relating to the quest.

**Review by KartulUdus — APPROVED** [review]
sweet, seems that rewards and conditions jumped back as I process them 
✅

**Review by kvangent — APPROVED** [review]

**Comment by kvangent** [comment]
@pgo-webhook-collaborators/realdevicemap and @pgo-webhook-collaborators/map-a-droid - Still waiting for reviews from your teams :)

**Comment by sn0opy** [comment]
Well, thanks to @fosJoddie MAD ist already compatible with the latest Poracle version which in fact already uses the proposed schema.

Thus, we're fine with this proposal.

**Comment by kvangent** [comment]
@sn0opy If only for the sake of ceremony, it would be great if you would use the review feature to approve so we can have those lovely green checkmarks :)

**Review by sn0opy — APPROVED** [review]

**Comment by 123FLO321** [comment]
Looks good but the naming scheme is not fully consistent.

For example:
- `template` is fine because it applies to the pokestop but the object is quest
- `longitude` is fine because applies to the quest (and pokestop also)
- `pokestop_id` is fine because it applies to the pokestop but the object is quest
- `quest_type` here just `quest` would be more logical because object is quest

**Comment by kvangent** [comment]
> * `quest_type` here just `quest` would be more logical because object is quest

Are you suggesting that we should change `quest_type` to just `quest`? That feels like it would introduce ambiguity -- if the field is just `quest` it doesn't describe what aspect of the quest it's describing (and just `type` is unclear what kind of type it's referring too since type is overloaded in a lot of different places)

**Comment by 123FLO321** [comment]
Sorry i meant to write just `type`. 
And it’s not unclear since it’s a quest object. We do it for most other stuff like template. So we either rename the other stuff to e.g. template to quest_template or just quest_type to type to make it consistent

**Comment by kvangent** [comment]
> Sorry i meant to write just `type`.
> And it’s not unclear since it’s a quest object. We do it for most other stuff like template. So we either rename the other stuff to e.g. template to quest_template or just quest_type to type to make it consistent

@123FLO321 Sorry for the delay on following up on this.

I think that `type` is potentially more confusing because it is used in many contexts - there is the event type, the quest type, the reward type, the rewards monster type. This doesn't really apply to `template` since it's the only time it's used. However, I'm okay with `quest_template` - I think it's better to be too descriptive than not enough.

I do have some concerns with using `quest_latitude` and `quest_longitude` since those fields are also common to the other events. Would you be satisfied if we update `template` and leave `latitude` and `longitude` as is?

**Comment by 123FLO321** [comment]
It's quest.type so that should be pretty clear.
quest.quest_type is redundant which is fine but we don't do it for the other stuff.
We should have everything named using one convention.

**Comment by fosJoddie** [comment]
I just want to point out, in between this quest_type vs type discussion. That this isn't technically the QUEST type, it is in fact the task type. 

If anything it should be named just task or task_type imo. in the context of quest, type alone is pretty clear. But it's not the type of quest, it is the type of task.

tl;dr: a quest has a task, the task is a type.

**Comment by kvangent** [comment]
I think `task` (or maybe even `task_id` to be clear it's a int representing the kind of task) sounds great to me.

**Comment by kvangent** [comment]
Bump on this - is everyone good with `task_id` over `quest_type`?

**Comment by kvangent** [comment]
@LegitDongo lets go ahead with  `quest_type` -> `task_id` so we can get this merged.

**Comment by LegitDongo** [comment]
Sorry folks, been in the process of moving for a bit now, will back this up once I am out of the time sink.
